src/librustc_traits/normalize_erasing_regions.rs

```diff
@@ -0,0 +1,37 @@
+// Copyright 2014 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use rustc::traits::{Normalized, ObligationCause};
+use rustc::traits::query::NoSolution;
+use rustc::ty::{ParamEnvAnd, Ty, TyCtxt};
+use rustc::util::common::CellUsizeExt;
+
+crate fn normalize_ty_after_erasing_regions<'tcx>(
+    tcx: TyCtxt<'_, 'tcx, 'tcx>,
+    goal: ParamEnvAnd<'tcx, Ty<'tcx>>,
+) -> Ty<'tcx> {
+    let ParamEnvAnd { param_env, value } = goal;
+    tcx.sess.perf_stats.normalize_ty_after_erasing_regions.increment();
+    tcx.infer_ctxt().enter(|infcx| {
+        let cause = ObligationCause::dummy();
+        match infcx.at(&cause, param_env).normalize(&value) {
+            Ok(Normalized { value: normalized_value, obligations: _ }) => {
+                //                                   ^^^^^^^^^^^
```

Comment:
How certain is this? Can you assert that this holds, or would that be too expensive?

Reply:
Hmm. It's probably worth asserting, or maybe refactoring the types to prove it.